WordPress/gutenberg
# `TabPanel`: support manual tab activation

Closes #45390

## What?


Adds support for manual tab activation to the `TabPanel` component (see https://github.com/WordPress/gutenberg/issues/45390 and  [W3C docs](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/) for more info)

## Why?


New feature for the `TabPanel` component, for enhanced accessibility when necessary.

## How?


- Introduced a new prop `hasManualTabActivation` (name TBD)
- The component keeps behaving as before when the new prop is not specified (backwards-compat)
- Added unit tests for both automatic and manual tab activation (including a necessary mock to test keyboard interactions)

## Testing Instructions





- Load the default `TabPanel` example in Storybook
- Click on the first tab, then use arrow key to switch tab. Note how the currently shown tabpanel changes as the arrow keys are pressed
- Via Storybook controls, change the value of the `hasManualTabActivation` prop to `true`
- Click on the first tab, then use arrow keys to navigate through the list of tabs. Notice how focusing a tab doesn't cause the currently shown tab panel to change.
- Make sure that the tabpanel changes only when the space or enter keys are pressed while focusing a tab.

## Screenshots or screencast 

https://user-images.githubusercontent.com/1083581/203551184-5f6f76d9-fa39-4128-bc72-dfd88d2b3f75.mp4
